libc/dns/getaddrinfo.c

```diff
@@ -43,11 +46,26 @@ int getaddrinfo(const char *name, const char *service,
                 const struct addrinfo *hints, struct addrinfo **res) {
   int rc, port;
   const char *canon;
+  size_t protolen;
   struct addrinfo *ai;
+  char proto[32];
   port = 0;
+
   if (!name && !service) return EAI_NONAME;
-  if (service && (port = parseport(service)) == -1) return EAI_NONAME;
   if (!name && (hints->ai_flags & AI_CANONNAME)) return EAI_BADFLAGS;
+  if (service) {
+    if (hints->ai_socktype == SOCK_STREAM)
+      strncpy(proto, "tcp", sizeof(proto));
```

Comment:
strncpy() is a misunderstood function. I only know of one legitimate use case for it. Doing binary search with a big endian buffer. For example:

```c
static const struct People {
  char name[8];
  int age;
} kPeople[] = {
    {"alice", 29},  //
    {"bob", 42},    //
};

int GetAge(const char *name) {
  char k[8];
  int m, l, r;
  l = 0;
  r = ARRAYLEN(kPeople) - 1;
  strncpy(k, s, 8);
  while (l <= r) {
    m = (l + r) >> 1;
    if (READ64BE(kPeople[m].s) < READ64BE(k)) {
      l = m + 1;
    } else if (READ64BE(kPeople[m].s) > READ64BE(k)) {
      r = m - 1;
    } else {
      return kPeople[m].age;
    }
  }
  return -1;
}
```

In this case it would be fine to say strcpy() since we know proto's length.